File: Java/TowerOfHanoi.java
/*Problem Statement : 

Tower of Hanoi is a mathematical puzzle where we have three rods and n disks. The objective of the puzzle is to move the entire stack to another rod, 
obeying the following simple rules:
1) Only one disk can be moved at a time.
2) Each move consists of taking the upper disk from one of the stacks and placing it on top of another stack i.e. a disk can only be moved if it is
the uppermost disk on a stack.
3) No disk may be placed on top of a smaller disk.

Time complexity : O(2^n)

*/

public class TowerOfHanoi {
    public static void HanoiProblem(int n,char rodFrom,char middleRod,char rodTo)
    {
        if(n==1)
        {
            System.out.println("Transfer rod "+n+" from "+rodFrom+" to "+rodTo);
            return;
        }
         HanoiProblem(n-1,rodFrom,rodTo ,middleRod);
         System.out.println("Transfer rod "+n+" from "+rodFrom+" to "+rodTo);
         HanoiProblem(n-1,middleRod,rodFrom,rodTo);
    }
    public static void main(String x[])
    {
        HanoiProblem(3,'A','B','C');
    }
    
}

File: C++/Tower_of_Hanoi.cpp
#include <stdio.h>
#include <iostream>
using namespace std;

static int count=0;
int towerOfHanoi(int n,char a,char b,char c){
    {
        if (n == 1)
        {
            cout<<"Move Disk from " << a << " to " << c<<endl;
            count++;
        }
        else {
            towerOfHanoi(n - 1, a, c, b);
            towerOfHanoi(1, a, b, c);
            towerOfHanoi(n - 1, b, a, c);
            
        }
        return count; 
    }
}
int main(){
    int n;
    cout<<"Enter the number of disks: "<<endl;
    cin>>n;
    int a = towerOfHanoi(n,'A','B','C');
    cout<<"Total number of moves = "<<a;
    return 0;
}
